pkg/printer/printer.go

```diff
@@ -265,6 +267,99 @@ func (p *Printer) WriteProtoFlow(f v1.Flow) error {
 	return nil
 }
 
+// joinWithCutOff performs a strings.Join, but will omit elements if the
+// resulting string is longer than targetLen. The resulting string may be
+// longer than targetLen, as it will always print at least one element.
+// The number of omitted elements is appended to the resulting string as
+// " (and N more)".
+func joinWithCutOff(elems []string, sep string, targetLen int) string {
+	strLen := 0
+	end := len(elems)
+	for i, elem := range elems {
+		strLen += len(elem) + len(sep)
+		if strLen > targetLen && i > 0 {
+			end = i
+			break
+		}
+	}
+
+	joined := strings.Join(elems[:end], sep)
+	omitted := len(elems) - end
+	if omitted == 0 {
+		return joined
+	}
+
+	return fmt.Sprintf("%s (and %d more)", joined, omitted)
+}
+
+// WriteProtoNodeStatusEvent writes a node status event into the error stream
+func (p *Printer) WriteProtoNodeStatusEvent(r *observerpb.GetFlowsResponse) error {
+	s := r.GetNodeStatus()
+	if s == nil {
+		return errors.New("not a node status event")
+	}
+
+	if !p.opts.enableDebug {
+		switch s.GetStateChange() {
+		case relaypb.NodeState_NODE_ERROR, relaypb.NodeState_NODE_UNAVAILABLE:
+			break
+		default:
+			// skips informal messages in non-debug mode
+			return nil
+		}
+	}
+
+	switch p.opts.output {
+	case JSONOutput:
+		return json.NewEncoder(p.opts.werr).Encode(r)
+	case DictOutput:
+		// this is a bit crude, but in case stdout and stderr are interleaved,
+		// we want to make sure the separators are still printed to clearly
+		// separate flows from node events.
+		if p.line != 0 {
+			_, err := fmt.Fprintln(p.opts.w, "------------")
+			if err != nil {
+				return err
+			}
+		} else {
+			p.line++
+		}
+		nodeNames := joinWithCutOff(s.NodeNames, ", ", nodeNamesCutOff)
+		message := "N/A"
+		if m := s.GetMessage(); len(m) != 0 {
+			message = strconv.Quote(m)
+		}
+		_, err := fmt.Fprint(p.opts.werr,
+			"  TIMESTAMP: ", fmtTimestamp(r.GetTime()), newline,
+			"      STATE: ", s.StateChange.String(), newline,
+			"      NODES: ", nodeNames, newline,
+			"    MESSAGE: ", message, newline,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to write out node status: %v", err)
+		}
+	default:
```

Comment:
i think default is better used for an "unknown type error" in case another enum is added

Reply:
I enumerated the cases explicitly to be consistent with the Flow printer. So it will silently ignore node status for new enum variants. I would find it confusing one threw an error, but the other one didn't.